Fix references to variance, add clarity

madaan opened the pull request:
1. Correct the references to "variance". The variance has been
referred to as standard deviation in several places.

2. The variance of the products of A[i] and X will be 512, not 1. (From: _We can show mathematically that as long as the elements in a and the elements in x are independent, the mean is 0 and the std is 1._). 

- The issue thus is that *each* element of y is now sampled from an erratic distribution, not just one element. In my opinion, this is one of the bigger points that the notebook currently misses. 

3. Add a note on how the normalization will fix point 2) above.

4. Add proofs (math + empirical) for points 2 and 3.

stas00 commented:
@madaan, thank you for your PR, but it doesn't pass the tests. Please refer to:
https://docs.fast.ai/dev/develop.html#things-to-run-after-git-clone
https://docs.fast.ai/dev/develop.html#stripping-out-jupyter-notebooks
and submit an updated notebook. Thanks.

madaan commented:
> @madaan, thank you for your PR, but it doesn't pass the tests. Please refer to:
> https://docs.fast.ai/dev/develop.html#things-to-run-after-git-clone
> https://docs.fast.ai/dev/develop.html#stripping-out-jupyter-notebooks
> and push an updated notebook. Thanks.

Done, thanks.

stas00 commented:
> Correct the references to "variance". The variance has been referred to as standard deviation in several places.

Reading through the first changes, you made a few valid corrections of where this indeed happened, but you also replaced 'std' with 'variance' where it was meant to be 'std'. Note that the code calls `x.std` and so the notes related to that output and not variance.

stas00 commented:
admin note: just configured reviewnb for this repo, so now it should be easy to review nb diffs from here. but since the initial event already happened for this PR, here is the direct link:

https://app.reviewnb.com/fastai/fastai_docs/pull/59/files/

stas00 commented on dev_course/dl2/02b_initializing.ipynb:
```diff
@@ -6,7 +6,8 @@
    "metadata": {},
```
<p>This diff needs to be reverted.</p>

 _Reply via <a href='https://app.reviewnb.com/fastai/fastai_docs/pull/59/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

stas00 replied:
<p>I think the above 2 Jeremy will object to, since you're changing the code - i.e. it won't match the video. On the other hand, this part wasn't really shown, so it's probably OK.</p>

 _Reply via <a href='https://app.reviewnb.com/fastai/fastai_docs/pull/59/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

stas00 replied:
<p>same as above. but if you keep it as the original then your proof doesn't match.</p>

 _Reply via <a href='https://app.reviewnb.com/fastai/fastai_docs/pull/59/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

stas00 commented:
See my specific comments above.

This was an excellent improvement, @madaan.

But it'd be @jph00 's final call if he's OK with the major changes.

madaan commented:
Thanks for the comments, @stas00. I'll give this another pass and fix the issues that you've pointed out. I'll also go ahead and make these fixes to the ReLu section. If Jeremy thinks that these changes are too major to add, I suggest that we come up with a few bare minimum fixes. As it stands, the notebook has a few issues that may confuse the students.

jph00 commented:
<p>Thanks to you both for all this work! :) Note that the original notebook assumes a mean of zero, and then uses that to just track x^2. That keeps things much more simple. These changes remove that simplification which makes it much harder to follow. Also, the new approach is averaging variances, which doesn't work. Finally, there are some proofs here which aren't relevant to what we're trying to show in the notebook, and are rather intimidating to people without a background in such things.</p><p><br></p><p>So I'd rather just see a "sqrt" added where needed (if there is anywhere) - note though that where we're only showing that something doesn't have a unit std, this isn't needed, since sqrt(1)==1.</p>

madaan commented:
Thanks for your comments! I'll work on a version that adds less details. The proofs can certainly be taken out/added elsewhere (perhaps at the end of the notebook?). 

My biggest concern is the following paragraph:

> Now that looks very close to the dimension of our matrix 512. And that's no coincidence! When you compute y, you sum 512 product of one element of a by one element of x. So what's the mean and the standard deviation of such a product? We can show mathematically that as long as the elements in a and the elements in x are independent, **_the mean is 0 and the std is 1_**. This can also be seen experimentally:

This is not true. In fact, if this was true, we wouldn't need any weight initialization at all! 

> **_the mean is 0 and the std is 1_**. 

The mean would indeed be 0, but the std would be sqrt(512). 

Just to make it clear, suppose we have an output vector:

`y = [y_0 y_1 y_2,...,y_n]`

The current write-up indicates that each of the `y_i` terms will have a mean = 0, std = 1 AND `y `as a whole has a mean = 0 and std = 512. Whereas in fact, each of the `y_i `terms have a mean = 0 and std = 512, which leads to y has a whole having a mean = 0 and std = 512. 

If each of the `y_i` terms (product terms) had a 0 mean and 1 std, `y` overall would in fact have 0 mean and std = 1.  I think this may lead to a bigger conceptual misunderstanding and needs to be clarified.

stas00 commented:
So to match the code to Jeremy's first para, it should have the missing `sqrt` and then I added printing a third number (sqr/100)**2 which now can be seen as ~512.
```
mean,sqr = 0.,0.
for i in range(100):
    x = torch.randn(512)
    a = torch.randn(512, 512)
    y = a @ x
    mean += y.mean().item()
    sqr  += y.pow(2).mean().sqrt().item()
mean/100, sqr/100, (sqr/100)**2
# gives
(-0.013176171109080315, 22.689782314300537, 514.8262214703454)
```
but the following paragraph/code is invalid as you correctly pointed out.

stas00 commented:
> The proofs can certainly be taken out/added elsewhere (perhaps at the end of the notebook?).

If I understand Jeremy's intention correctly, we want as little math as possible in the teaching materials, because otherwise it looks intimidating to those w/o a strong academic background and they are likely to bail.

I'm not sure what would be the best recommendation here. Perhaps this could go into the notes as an appendix? https://forums.fast.ai/t/lesson-8-notes/41442

sgugger commented:
> > Now that looks very close to the dimension of our matrix 512. And that's no coincidence! When you compute y, you sum 512 product of one element of a by one element of x. So what's the mean and the standard deviation of such a product? We can show mathematically that as long as the elements in a and the elements in x are independent, **_the mean is 0 and the std is 1_**. This can also be seen experimentally:
> 
> This is not true. In fact, if this was true, we wouldn't need any weight initialization at all!

I'm sorry but this is exactly what is shown in the demonstration at the end. the a_ij * x_j all have a mean of 0 and a variance of 1, thus summing 512 of them gives a mean of 0 and a variance of 512.

It's very kind of you to add the mathematical proof, but the notebook was supposed to show the intuition behind it without scaring anyone with math (that's word for word what Jeremy asked me to do).  The confusion between variance and std needs to be cleared, I agree, and maybe some wording need to be bettered if it's unclear in the current state, but the actual experiments and the general outline should stay the same. 
I also think the math should be done in a separate notebook that we can write from this one, and developed as the proof is outlined, but the current notebook should stay clear of it and just outline the intuition of the proof.

madaan commented:
Thanks team for taking out the time to look into this in detail, and for your comments. 

I was referring to the statement:

> And that's no coincidence! When you compute y, you sum 512 product of one element of a by one element of x. So what's the mean and the standard deviation of **such a product?** We can show mathematically that as long as the elements in a and the elements in x are independent, the mean is 0 and the std is 1.

I guess the confusion is regarding the statement "such a product". If we're taking about the product of one x and one a, this holds. However, if the statement is talking about the calculation of y in general, it's not. 


I've moved the proofs to the end and have added some consistency to the experiments.

madaan commented:
> So to match the code to Jeremy's first para, it should have the missing `sqrt` and then I added printing a third number (sqr/100)**2 which now can be seen as ~512.
> 
> ```
> mean,sqr = 0.,0.
> for i in range(100):
>     x = torch.randn(512)
>     a = torch.randn(512, 512)
>     y = a @ x
>     mean += y.mean().item()
>     sqr  += y.pow(2).mean().sqrt().item()
> mean/100, sqr/100, (sqr/100)**2
> # gives
> (-0.013176171109080315, 22.689782314300537, 514.8262214703454)
> ```
> so I think that more or less matches the para explaining it.
> 
> The next code section seems to try to show that variance is 1, and so then sqrt of 1 is 1, so the std is 1.
> 
> ```
> mean,sqr = 0.,0.
> for i in range(10000):
>     x = torch.randn(1)
>     a = torch.randn(1)
>     y = a*x
>     mean += y.item()
>     sqr  += y.pow(2).item()
> mean/10000,sqr/10000
> # gives
> (0.009260928024939597, 1.0256406533188511)
> ```

Sounds like a good idea, I've added both var and std in the latest commit.

sgugger commented:
Ok, so there are three different problems here: one is the fix std vs variance, the other is the sketch of the proof in English and the last one is the math. Having them all in one PR makes it too complex to adjust one of them so I'll close this then fix the std vs variance problem.

Then if the sketch of the proof needs rewording, please feel free to suggest a PR that just addresses that and we can discuss more easily.

As for the math, I just checked with Jeremy and we feel it really should be in another notebook to not scare anyone, so I think the best is for you to host this notebook on a gist and we can link to it from our 02b and in the forums, for anyone interested in reading more.

In any case, thanks for pointing out the issues and trying to fix them, but we need to break this down in separate smaller increments.

madaan commented:
Sounds good @sgugger, thanks. Is it okay if I host the proof etc. on my blog instead and then share a link to the blog on the forums?

stas00 commented:
@madaan, the other suggestion I made earlier is to make your proof into a standalone appendix and perhaps it can be added to the notes for lesson 9 (or linked to from the notes and resources of lesson 9). I don't see a thread for notes for lesson 9 yet. So perhaps for now just add a link to in the first post (wiki) of https://forums.fast.ai/t/lesson-9-discussion-wiki-2019/41969

sgugger commented:
> Sounds good @sgugger, thanks. Is it okay if I host the proof etc. on my blog instead and then share a link to the blog on the forums?

I have absolutely no problem with that!
